### PR TITLE
[Ready? TM Ready] Diabolical resurrection resurrection

### PR DESCRIPTION
**Pull request by alex-gh**
A deconflicted version of #5890

Most of the edits @tigercat2000 requested are in.

Known problems:
- Dance floor is summoned airless the first time (only on shuttle).
- (Revival contracts are broken) Could not reproduce - works for me. 

TODO:
- Make people, who sold their souls appear in check antagonists.

-Take a look at https://github.com/tgstation/tgstation/tree/master/code/modules/antagonists/devil
https://github.com/tgstation/tgstation/pull/26815/files

**Changelog:**
:cl: lordPidey
rscadd: Infernal devils have been seen onboard our spacestations, offering great boons in exchange for your soul.
rscadd: Employees are reminded that their soul already belongs to Nanotrasen. If you have sold your soul in error, a lawyer or head of personnel can help return your soul to Nanotrasen by hitting you with your employment contract.
rscadd: Nanotrasen headquarters will be bluespacing employment contracts into the IA's office filing cabinet when a new arrival reaches the station. It is recommended that the lawyer create copies of some of these for safe keeping.
rscadd: Due to the recent infernal incursions, the station library has been equipped with a Codex Gigas to help research infernal weaknesses. Please note that reading this book may have unintended side effects. Also note, you must spell the devil's name exactly, as there are countless demons with similar names.
rscadd: When a devil dies, if the proper banishment ritual is not performed on it's remains, the devil will revive itself at the cost of some of it's power. The banishment ritual is described in the Codex Gigas.
rscadd: When a demon gains enough souls, It's form will mutate to a more demonic looking form. The Arch-demon form is known to be on par with an ascended shadowling in power.
add: Added Devil agent gamemode, where multiple devils are each trying to buy more souls than the next in line.
add: If you've already sold your soul, you can sell it again to a different devil. You can even go back and forth for INFINITE POWER.
/:cl:

### CODE REVIEW

**Comment by KasparoVy** [comment]
How much deeper can we go

8th layer, all the way down :smiling_imp:

**Comment by Injazz** [comment]
sadly, devils doesn't work as intended
the urge for valids is what ruined it

**Comment by Quantum-M** [comment]
The issue with devils isn't the devil themselves, although making security actually depend on the Codex Gygax to banish the devil instead of just bashing it to death till it's gone would be nice. The big issue is the contracts themselves, which just encourages greytide behavior in general.

**Comment by Kyep** [comment]
Problems:
- Nothing shows up in check antagonists.
- Nothing shows up on antaghud.
- Roles that shouldn't be eligible to be devils, such as silicons, get chosen.
- I have no idea what the difference between 'devil' and 'devil agents' is.

File: code/modules/paperwork/contract.dm [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+		return FALSE
+
+	if(user.mind != target)
+		to_chat(user,"<span class='notice'>Your signature simply slides off the sheet, it seems this contract is not meant for you to sign.</span>")
+		return FALSE
+
+	if(!user.mind.hasSoul)
+		to_chat(user, "<span class='notice'>You lack a soul to sign away!</span>")
+
+	if(user.mind.soulOwner != owner)
```
Comment (Kyep) [inline code/modules/paperwork/contract.dm]: Double-check this.

File: code/game/objects/items/weapons/defib.dm [hunk 2]
```diff
@@ -512,15 +512,15 @@
...
+				if(H.stat == DEAD)
 					var/health = H.health
 					M.visible_message("<span class='warning'>[M]'s body convulses a bit.")
 					playsound(get_turf(src), "bodyfall", 50, 1)
 					playsound(get_turf(src), 'sound/machines/defib_zap.ogg', 50, 1, -1)
 					for(var/obj/item/organ/external/O in H.bodyparts)
 						total_brute	+= O.brute_dam
 						total_burn	+= O.burn_dam
-					if(total_burn <= 180 && total_brute <= 180 && !H.suiciding && !ghost && tplus < tlimit && !(NOCLONE in H.mutations))
+					if(total_burn <= 180 && total_brute <= 180 && !H.suiciding && !ghost && tplus < tlimit && !(NOCLONE in H.mutations) && !(H.mind && H.mind.is_revivable()))
```
Comment (Ty-Omaha) [inline code/game/objects/items/weapons/defib.dm]: this is inverted (and NOT mind and NOT revivable)

**Comment by taukausanake** [comment]
> encourages greytide behavior

I would imagine that's part of the appeal of this gamemode as the devils can't do much of anything in early round. It's like, "Hey I'm the devil and I'm here to fuck you all over by helping you fuck each other over here have a free thing to ensue chaos with and keep people not focused on me."

**Comment by lordpidey** [comment]
Original author here.

> Roles that shouldn't be eligible to be devils, such as cyborgs, and members of security, get chosen.

Cyborgs are intentional, and even have code to insert their devil obligations as laws. (At least on TG).  I'm not as familiar with paradise's community, if you want to remove these, that's okay.

> I have no idea what the difference between 'devil' and 'devil agents' is.

Same as between traitor and double agents.  In devil agents, each devil has an objective to buy more souls than another specific devil.

> Nothing shows up on antaghud.

> De-deviling someone doesn't remove their HUD, apparently.

On TGstation at least, devils aren't supposed to see the devil huds, the devil hud is for admins to see.


>TODO:
>Make people, who sold their souls appear in check antagonists.

Keep in mind soul sellers are NOT antagonists.  They're just unrevivable crewmembers with superpowers.

File: code/game/gamemodes/devil/devilinfo.dm [hunk 3]
```diff
@@ -383,6 +383,8 @@ var/global/list/lawlorify = list (
 		to_chat(owner.current, "<span class='danger'>You seem to have resurrected without your hellish powers.</span>")
 
 /datum/devilinfo/proc/check_banishment(mob/living/body)
+	if(QDELETED(body))	//Allow ressurection if the body has been destroyed
+		return 0
```
Comment (lordpidey) [inline code/game/gamemodes/devil/devilinfo.dm]: No, this will break the banish ritual where you are SUPPOSED to destroy the body, take a look at line 415

Comment (alex-gh) [reply code/game/gamemodes/devil/devilinfo.dm]: Good catch, thanks.

**Comment by alex-gh** [comment]
> Original author here.
> 
> > Roles that shouldn't be eligible to be devils, such as cyborgs, and members of security, get chosen.
> 
> Cyborgs are intentional, and even have code to insert their devil obligations as laws. (At least on TG). I'm not as familiar with paradise's community, if you want to remove these, that's okay.

Having Cyborgs be round start antags would be rather unusual.
Also there is a matter of law sync, where we would end up with borgs synced to the AI but not sharing its laws.

**Comment by lordpidey** [comment]
> > Cyborgs are intentional, and even have code to insert their devil obligations as laws. (At least on TG). I'm not as familiar with paradise's community, if you want to remove these, that's okay.
> 
> Having Cyborgs be round start antags would be rather unusual.
> Also there is a matter of law sync, where we would end up with borgs synced to the AI but not sharing its laws.

The intent is that the updated laws ONLY show up for the cyborg, who is free to do whatever, and from the AI's point of view, it's just a normal cyborg that's synched with them.  If the borg is synched, and the AI's laws change, the cyborg's laws will change too, but the cyborg will keep it's law 666.

**Comment by alex-gh** [comment]
> > Having Cyborgs be round start antags would be rather unusual.
> > Also there is a matter of law sync, where we would end up with borgs synced to the AI but not sharing its laws.
> 
> The intent is that the updated laws ONLY show up for the cyborg, who is free to do whatever, and from the AI's point of view, it's just a normal cyborg that's synched with them. If the borg is synched, and the AI's laws change, the cyborg's laws will change too, but the cyborg will keep it's law 666.

Sure, there are no problems with that mechanically. It is just inconsistent with how silicon mechanics work now (Either borgs have law sync and same laws as AI, or borg has no law sync and possibly different laws).

Also I just don't know if having a silicon spell caster is a good idea...

I'll bring it up to staff again, see what they think.

**Pull request merged**